From d5b74ea0ebf7d501e7520db47ea2fbd7dfabf8b9 Mon Sep 17 00:00:00 2001
From: Weiguang Ruan <Weiguang.ruan@amlogic.com>
Date: Fri, 23 Nov 2018 17:08:26 +0800
Subject: [PATCH] wifi: add qca9377 wifi support in S42wifi/multiwiifi/web_ui [1/3]

PD#SH-23

Problem:
S42wifi/multiwifi/wei_ui do not support qca9377

Solution:
add support

Verify:
S400

Change-Id: I7a7671bf5c66b322bfed0841f93ef2a3a5bdde07
Signed-off-by: Weiguang Ruan <Weiguang.ruan@amlogic.com>
---

diff --git a/board/amlogic/common/rootfs/rootfs-49-audio/etc/init.d/S42wifi b/board/amlogic/common/rootfs/rootfs-49-audio/etc/init.d/S42wifi
index 1af2014..3b795bf 100755
--- a/board/amlogic/common/rootfs/rootfs-49-audio/etc/init.d/S42wifi
+++ b/board/amlogic/common/rootfs/rootfs-49-audio/etc/init.d/S42wifi
@@ -182,7 +182,8 @@
 	0x02d0)
 		wifi_setup both
 		;;
-	0x024c)
+	#we use qca9377 same with rtk temperary
+	0x024c|0x0271)
 		RTK_WIFI_FLAG="TRUE"
 		wifi_rtk
 		;;
@@ -206,8 +207,7 @@
 		$MULTI_WIFI station 0
         $POWERCTL 0
         ;;
-    netready|netup|netdown|netchange) ;;
-    restart|reload)
+    restart)
         killall hostapd
         killall wpa_supplicant
         killall dnsmasq
diff --git a/package/aml_util/aml_util.mk b/package/aml_util/aml_util.mk
index c3881a2..d1a4088 100755
--- a/package/aml_util/aml_util.mk
+++ b/package/aml_util/aml_util.mk
@@ -42,6 +42,10 @@
 	AML_UTIL_PRIV_FLAGS += -DMRVL_MODULES_PATH=/lib/modules/$(LINUX_VERSION_PROBED)/kernel/mrvl
 endif
 
+ifneq ($(BR2_PACKAGE_QUALCOMM_WIFI),)
+	AML_UTIL_PRIV_FLAGS += -DQCA_MODULES_PATH=/lib/modules/$(LINUX_VERSION_PROBED)/kernel/qualcomm/wifi
+endif
+
 
 define AML_UTIL_BUILD_CMDS
 	$(TARGET_CONFIGURE_OPTS) $(MAKE) -C $(@D) CFLAGS="$(AML_UTIL_PRIV_FLAGS)" all
diff --git a/package/aml_util/src/multi_wifi_load_driver.c b/package/aml_util/src/multi_wifi_load_driver.c
index 4d8ec79..66245e7 100644
--- a/package/aml_util/src/multi_wifi_load_driver.c
+++ b/package/aml_util/src/multi_wifi_load_driver.c
@@ -48,9 +48,16 @@
 #define MRVL_KO_PATH     DEFAULT_CONFIG_PATH
 #endif
 
+#ifdef QCA_MODULES_PATH
+#define QCA_KO_PATH     XSTR(QCA_MODULES_PATH)
+#else
+#define QCA_KO_PATH     DEFAULT_CONFIG_PATH
+#endif
+
 #define MODULE_ARG_FIRMWARE     0
 #define MODULE_ARG_IFNAME       1
 #define MODULE_ARG_STACFG       2
+#define MODULE_ARG_OTHER        3
 
 #if defined(__NR_finit_module)
 # define finit_module(fd, uargs, flags) syscall(__NR_finit_module, fd, uargs, flags)
@@ -64,6 +71,7 @@
     const char *ifname;
     const char *if2name;
     const char *stacfgpath;
+    const char *arg;
 } module_arg;
 
 typedef struct load_info {
@@ -304,6 +312,18 @@
 		"sd8xxx",
 		"sd8xxx_8987.ko",
 		"cal_data_cfg=none"
+	},
+	{
+		"0701",
+		"wlan",
+		"wlan.ko",
+		QCA_KO_PATH,
+		.wifi_module_arg = {
+			.arg_type   = MODULE_ARG_OTHER,
+			.arg        = "country_code=CN",
+		},
+		"qca9377",
+		0x0,
 	}
 };
 
@@ -324,10 +344,14 @@
     case MODULE_ARG_STACFG: {
 	sprintf(str, "stacfgpath=%s", arg->stacfgpath);
 	break;
-    }
-    default:
+	}
+	case MODULE_ARG_OTHER: {
+	sprintf(str, "%s", arg->arg);
 	break;
-    }
+	}
+	default:
+	break;
+	}
 }
 
 static char file_name[100] = {'\0'};
diff --git a/package/web_ui_wifi/src/cgi-bin/scripts/wifi/wifi_tool.sh b/package/web_ui_wifi/src/cgi-bin/scripts/wifi/wifi_tool.sh
index d1b0bee..e649224 100755
--- a/package/web_ui_wifi/src/cgi-bin/scripts/wifi/wifi_tool.sh
+++ b/package/web_ui_wifi/src/cgi-bin/scripts/wifi/wifi_tool.sh
@@ -33,7 +33,7 @@
 	wifi_chip_id_vendor="/sys/bus/mmc/devices/sdio:0001/sdio:0001:1/vendor"
 	wifi_chip_id=`cat ${wifi_chip_id_vendor}`
 	case "${wifi_chip_id}" in
-		0x024c)
+		0x024c|0x0271)
 		RTK_WIFI_FLAG="TRUE"
 		;;
 	esac
@@ -54,13 +54,7 @@
 echo "Stopp prv dhcpcd first"
 start-stop-daemon -K -o -p $PIDFILE4 2> /dev/null
 sleep 1
-echo "delete prv br0"
-ifconfig | grep br0 > /dev/null
-if [ $? -eq 0 ];then
-	ifconfig br0 down > /dev/null
-	brctl delbr br0
-fi
-sleep 1
+${MULTI_WIFI} station 0
 }
 
 #uninstall rtk driver
